rpcs3/rpcs3qt/syntax_highlighter.cpp

```diff
@@ -0,0 +1,122 @@
+#include "syntax_highlighter.h"
+
+syntax_highlighter::syntax_highlighter(QTextDocument *parent) : QSyntaxHighlighter(parent)
+{
+}
+
+void syntax_highlighter::AddSimpleRule(SimpleRule rule)
```

Comment:
why pass by value?

Reply:
I think the removeAll didn't work with const ref

Reply:
hmm normal ref is fine? xD

Reply:
@Megamouse 
Yeah, you are right. didn't notice that.

But I think it's better to rewrite that loop into:
```cpp
rule.expressions.erase(std::remove_if(rule.expressions.begin(), rule.expressions.end(), [](const auto& expression){
    return !expression.isValid() || expression.pattern().isEmpty();
}), rule.expressions.end());
```

Since otherwise you'd be doing `O(n^2)` removals, if I'm not mistaken (since afaik `removeAll` scans the whole list on each call).

Reply:
btw i ended up changing it again to a reversed loop

Reply:
@Megamouse hm, why? It looks wrong. You always removeLast element but only check for the `i`th one. So if the last one okay but the one before him is not, you'll still remove the last.
Why doing unnecessary loop, if you need to add another condition you can just do:
```cpp
rule.expressions.erase(std::remove_if(rule.expressions.begin(), rule.expressions.end(), [&](const auto &e){
	return IsInvalidExpression(e) || e.captureCount() > 1;
}), rule.expressions.end());
```

Reply:
oh man I'm super tired. might be the reason.... but what's so wrong with a simple loop.
Is it slower than the erase?

Reply:
```cpp
for (int i = rule.expressions.length() - 1; i > 0; i--)
{
	if (IsInvalidExpression(rule.expressions[i]) || rule.expressions[i].captureCount() > 1)
	{
		rule.expressions.removeLast();
	}
}
```

What element would be removed if the `if` condition is `false` for the `i = rule.expressions.length() - 1`, but `true` for `rule.expressions.length() - 2`? Also, this loop ignores the `0` element (not that it matters as it's broken as it is).

Basically it won't remove the element for each the predicate is `true`, it'll remove `k` last elements if there are `k` elements in the `expressions` for each it's true.

Reply:
Oh. I didn't even remove the erase.... It was meant to be replaced. 🥇 
Of course rule.expressions[i] shall be removed. My bad. No excuse for the ignored 0 either.

But still: is the loop actually slower than erase(remove_if)?

Reply:
nvm I read the wiki page. gonna do it the common way

Reply:
@Megamouse >But still: is the loop actually slower than erase(remove_if)?

In the worst case - yes. Since simple remove/erase is `O(n)` for vectors so calling it about `n` times in a loop is worse, than just a linear `erase(remove_if)`. Also, loop is just harder to read/maintain (look at the error you've made for example).